cmake/all.cmake

```diff
@@ -165,6 +169,9 @@ message(STATUS "catkin ${catkin_VERSION}")
 # ensure that no current package name is set
 unset(_CATKIN_CURRENT_PACKAGE)
 
+# tools/libraries.cmake
+configure_shared_library_build_settings()
```

Comment:
Doesn't adding this change the current behavior? Before the option was not set which make `add_library` default to `STATIC`. With this line being added which basically runs this: https://github.com/ros/catkin/blob/964f1bf385f356c8345677928801947d497a161e/cmake/tools/libraries.cmake#L11-L13

the default behavior changes to `SHARED`.

Reply:
the BUILD_SHARED_LIBS is turned on in catkin/cmake/platform/windows.cmake, so it's ON by default (even on *nix).  In addition, configure_shared_library_build_settings() is called by catkin/cmake/catkin_workspace.cmake, too.  Which means BUILD_SHARED_LIBS is turned on by default in toplevel.cmake.  I'm not sure if add_library is default to STATIC, from my observation, it's default to SHARED.
  Do we have a consensus about the default value of this switch?  We need to nail down the default value and then change the code to make it consistent.

Reply:
The `toplevel.cmake` / `catkin_workspace` is only being used by `catkin_make`. All other build tools (`catkin_make_isolated`, `catkin-tools`, `colcon`) don't use a workspace-level CMake file but invoke CMake for each package separately.

But I think you are right that `BUILD_SHARED_LIBS` is turned on by default. The reason is that both (one would already be sufficient) of the following files are included in the `all.cmake` file ([platform/windows.cmake](https://github.com/ros/catkin/blob/964f1bf385f356c8345677928801947d497a161e/cmake/all.cmake#L148), [tools/libraries.cmake](https://github.com/ros/catkin/blob/964f1bf385f356c8345677928801947d497a161e/cmake/all.cmake#L154)). Both declare the option and ensure that it is initialized with `on` if not already set.

Can you clarify why explicitly calling `configure_shared_library_build_settings()` here is necessary for making `nmake` work on Windows?

Reply:
For Windows, all shared libraries are built into Dll files and need to specify the storage-class attributes to indicate if a function/data is exported (the dll provides the function/data) or imported (the dll consumes the function/data).  We use macro to decide what attribute to use when compile a component (see https://github.com/ros/ros_comm/blob/melodic-devel/tools/rosbag_storage/include/rosbag/macros.h for example).  The macro checks the flag ROS_BUILD_SHARED_LIBS to decide the storage-class attribute.  configure_shared_library_build_settings() is where to decide whether to set the ROS_BUILD_SHARED_LIBS.

Reply:
Thanks for clarifying - I forgot that this is what we use `ROS_BUILD_SHARED_LIBS` for.